src/build/plan.rs

```diff
@@ -194,32 +195,36 @@ impl Plan {
             })
             .collect();
 
-        for modified in files {
-            if let Some(unit) = build_scripts.get(modified.as_ref()) {
+        for modified in files.iter().map(|x| x.as_ref()) {
+            if let Some(unit) = build_scripts.get(modified) {
                 result.insert(unit.clone());
             } else {
-                // Not a build script, so we associate a dirty package with a
-                // dirty file by finding longest (most specified) path prefix
-                let unit = other_targets.iter().max_by_key(|&(_, src_dir)| {
-                    if !modified.as_ref().starts_with(src_dir) {
-                        return 0;
-                    }
-                    modified
-                        .as_ref()
-                        .components()
-                        .zip(src_dir.components())
-                        .take_while(|&(a, b)| a == b)
+                // Not a build script, so we associate a dirty file with a
+                // package by finding longest (most specified) path prefix.
+                let matching_prefix_components = |a: &Path, b: &Path| -> usize {
+                    assert!(a.is_absolute() && b.is_absolute());
+                    a.components().zip(b.components())
+                        .skip(1) // Skip RootDir
+                        .take_while(|&(x, y)| x == y)
                         .count()
-                });
-                match unit {
-                    None => trace!(
-                        "Modified file {:?} doesn't correspond to any package!",
-                        modified
-                    ),
-                    Some(unit) => {
-                        result.insert(unit.0.clone());
-                    }
                 };
+                // Since a package can correspond to many units (e.g. compiled
+                // as a regular binary or a test harness for unit tests), we
+                // collect every unit having the longest path prefix.
+                let max_matching_prefix = other_targets.values()
+                    .map(|src_dir| matching_prefix_components(modified, src_dir))
+                    .max().unwrap();
```

Comment:
It's not obvious that this `unwrap` will never trip :)

Reply:
Nice catch, thanks!